techmatters/serverless
# Support status-based messages to be included in operatingHours response

## Description
This PR modifies the following functions:
- `functions/operatingHours.ts`: now supports two new optional parameters in the request: `includeMessageTextInResponse` and `language`. If `includeMessageTextInResponse` is not provided, the function will continue to work as before (prevent breaking changes). If the function is called with `includeMessageTextInResponse` = `true`, the changes on this PR will apply: the response won't be a plain `string` representing the status (`open`, `closed`, `holiday`), but an object of the shape `{ status: string; message: string }`. The message will indicate the reason (if closed due to shifts or holidays). If `language` is provided, we'll attempt to retrieve the translated version, defaulting to `en-US` if not.
- `functions/sendSystemMessage.ts`: now supports providing `taskSid` or `channelSid` as the destination for the message. Exactly one must be provided.
- Adds `functions/sendMessageAndRunJanitor.protected.ts` which is just a wrapper around `sendSystemMessage`, that does not asks for Token validation, but being marked as `.protected` will only accept requests with a valid [X-Twilio-Signature header](https://www.twilio.com/docs/usage/security). This is required so we can reuse the `sendSystemMessage` functionality from the context of a Studio Flow, and there's no `Token` to include in the request. This function also:
  - Removes the `'studio'` type webhooks from the channel to avoid triggering a second execution. This is needed to fix the [bottom section in this description](https://github.com/techmatters/serverless/pull/409#note:~:text=I%27m%20seeing%20an,a%20bit%20hacky.).
  - Invokes the chat channel janitor to perform a cleanup on the channel (not sure if required but won't hurt deactivating the maybe-stale channel).

### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/browse/CHI-1635)

### Verification steps
- Make sure this branch is deployed to Aselo Development.
- Go to Twilio Console -> Functions -> Serverless and manually remove (or set to empty) the environment variable `DISABLE_OPERATING_HOURS_CHECK` if it is present (this way we'll get operating hours checks).
- Make sure that the Messaging Studio Flow is wired like in the picture above.
- Sending messages from Facebook and Instagram should always return "closed due to shift" (because of the shape of `aselo-dev` operating info asset). Messages from webchat should succeed.
- Following the validation steps of https://github.com/techmatters/webchat/pull/184 might also help :).
- Once done testing, either turn set `DISABLE_OPERATING_HOURS_CHECK` back to true or re-deploy.


Studio Flow:

![Screenshot from 2023-02-02 17-08-58](https://user-images.githubusercontent.com/15805319/216442172-34d0ae04-3930-4585-a423-7013bfaee207.png)

The updates in the Studio Flow consist of the following (transitions):
- Determine office (`split office` widget).
- Save the variables of the current office (`set X office`).
- Run function `open hours` with
  - `channel`: `{{flow.variables.channel}}`.
  - `office`: `{{flow.variables.channel}}`.
  - `includeMessageTextInResponse`: true :bangbang: This is new.
  - `language` (optional): the language from where it is stored (pre engagement for webchat, maybe a bot's memory for other channels?
- Determine the response from above using a split widget based on `widgets.openHours_office.parsed.status`. If
  - Contains `open` -> follow normal flow execution.
  - No condition matched means that it is either `closed` or `holiday` -> Run function `/sendMessageAndRunJanitor` with:
    - `channelSid`: `{{flow.channel.address}}`.
    - `message`: `{{widgets.openHours_office.parsed.message}}`.

### Note:
I'm seeing an unexpected behavior, the function to send the message is being called multiple times, resulting in repeated messages being sent to the channels. This is occurring for custom and Twilio channels. I have a support ticket asking for this behavior, but if you have some idea the help is very appreciated.
Update on this: looks like the message being sent by the `system` to the active channel, having a different `from` address is causing a second Studio Flow execution. This is likely the reason why Twilio does not includes the `xTwilioWebhookEnabled = true` for the "send message widget" :) I'm hoping to get some assistance from Support here. A potential idea to prevent this behavior would be to catch every incoming message at the beginning of the Studio Flows, and just "do nothing" if the message is comming from `system`, but feels a bit hacky.
